File: code/math_utils/geo.py
"""
geo.py
PURPOSE: Convert between different ways of describing locations on Earth.
"""

import math
import numpy as np
from typing import Tuple

# WGS84 ellipsoid parameters
WGS84_A = 6378137.0  # semi-major axis (meters)
WGS84_B = 6356752.314245  # semi-minor axis (meters)
WGS84_E2 = 1.0 - (WGS84_B ** 2) / (WGS84_A ** 2)  # first eccentricity squared


def wgs84_to_ecef(lat: float, lon: float, alt: float) -> Tuple[float, float, float]:
    """
    Convert GPS coordinates to Earth-Centered Earth-Fixed (ECEF).
    
    Args:
        lat: Latitude in degrees
        lon: Longitude in degrees
        alt: Altitude in meters above WGS84 ellipsoid
    
    Returns:
        Tuple of (X, Y, Z) in meters
    """
    lat_rad = math.radians(lat)
    lon_rad = math.radians(lon)
    
    sin_lat = math.sin(lat_rad)
    cos_lat = math.cos(lat_rad)
    sin_lon = math.sin(lon_rad)
    cos_lon = math.cos(lon_rad)
    
    # Radius of curvature in the prime vertical
    N = WGS84_A / math.sqrt(1.0 - WGS84_E2 * sin_lat * sin_lat)
    
    X = (N + alt) * cos_lat * cos_lon
    Y = (N + alt) * cos_lat * sin_lon
    Z = (N * (1.0 - WGS84_E2) + alt) * sin_lat
    
    return (X, Y, Z)


def ecef_to_enu(
    x: float, y: float, z: float,
    ref_lat: float, ref_lon: float, ref_alt: float
) -> Tuple[float, float, float]:
    """
    Convert ECEF to local ENU (East-North-Up) coordinates.
    
    Args:
        x, y, z: ECEF coordinates in meters
        ref_lat, ref_lon, ref_alt: Reference point (origin) in WGS84
    
    Returns:
        Tuple of (E, N, U) in meters relative to reference
    """
    # Get reference point in ECEF
    ref_x, ref_y, ref_z = wgs84_to_ecef(ref_lat, ref_lon, ref_alt)
    
    # Difference vector
    dx = x - ref_x
    dy = y - ref_y
    dz = z - ref_z
    
    # Rotation angles
    lat_rad = math.radians(ref_lat)
    lon_rad = math.radians(ref_lon)
    
    sin_lat = math.sin(lat_rad)
    cos_lat = math.cos(lat_rad)
    sin_lon = math.sin(lon_rad)
    cos_lon = math.cos(lon_rad)
    
    # ENU transformation
    E = -sin_lon * dx + cos_lon * dy
    N = -sin_lat * cos_lon * dx - sin_lat * sin_lon * dy + cos_lat * dz
    U = cos_lat * cos_lon * dx + cos_lat * sin_lon * dy + sin_lat * dz
    
    return (E, N, U)


def wgs84_to_enu(
    lat: float, lon: float, alt: float,
    ref_lat: float, ref_lon: float, ref_alt: float
) -> Tuple[float, float, float]:
    """
    Convert GPS directly to local ENU coordinates.
    
    Args:
        lat, lon, alt: Target point in WGS84
        ref_lat, ref_lon, ref_alt: Reference point (origin) in WGS84
    
    Returns:
        Tuple of (E, N, U) in meters
    """
    x, y, z = wgs84_to_ecef(lat, lon, alt)
    return ecef_to_enu(x, y, z, ref_lat, ref_lon, ref_alt)


def enu_to_wgs84(
    e: float, n: float, u: float,
    ref_lat: float, ref_lon: float, ref_alt: float
) -> Tuple[float, float, float]:
    """
    Convert local ENU coordinates back to WGS84.
    
    Args:
        e, n, u: Local ENU coordinates in meters
        ref_lat, ref_lon, ref_alt: Reference point (origin) in WGS84
    
    Returns:
        Tuple of (lat, lon, alt) in degrees/meters
    """
    # Get reference point in ECEF
    ref_x, ref_y, ref_z = wgs84_to_ecef(ref_lat, ref_lon, ref_alt)
    
    # Rotation angles
    lat_rad = math.radians(ref_lat)
    lon_rad = math.radians(ref_lon)
    
    sin_lat = math.sin(lat_rad)
    cos_lat = math.cos(lat_rad)
    sin_lon = math.sin(lon_rad)
    cos_lon = math.cos(lon_rad)
    
    # Inverse ENU transformation (to ECEF delta)
    dx = -sin_lon * e - sin_lat * cos_lon * n + cos_lat * cos_lon * u
    dy = cos_lon * e - sin_lat * sin_lon * n + cos_lat * sin_lon * u
    dz = cos_lat * n + sin_lat * u
    
    # ECEF position
    x = ref_x + dx
    y = ref_y + dy
    z = ref_z + dz
    
    # Convert ECEF back to WGS84 (iterative)
    return ecef_to_wgs84(x, y, z)


def ecef_to_wgs84(x: float, y: float, z: float) -> Tuple[float, float, float]:
    """
    Convert ECEF to WGS84 (iterative method).
    
    Args:
        x, y, z: ECEF coordinates in meters
    
    Returns:
        Tuple of (lat, lon, alt) in degrees/meters
    """
    lon = math.atan2(y, x)
    
    # Iterative solution for latitude and altitude
    p = math.sqrt(x * x + y * y)
    lat = math.atan2(z, p * (1.0 - WGS84_E2))
    
    for _ in range(10):  # Usually converges in 2-3 iterations
        sin_lat = math.sin(lat)
        N = WGS84_A / math.sqrt(1.0 - WGS84_E2 * sin_lat * sin_lat)
        lat = math.atan2(z + WGS84_E2 * N * sin_lat, p)
    
    sin_lat = math.sin(lat)
    N = WGS84_A / math.sqrt(1.0 - WGS84_E2 * sin_lat * sin_lat)
    alt = p / math.cos(lat) - N
    
    return (math.degrees(lat), math.degrees(lon), alt)


def calculate_look_angles(
    observer_e: float, observer_n: float, observer_u: float,
    target_e: float, target_n: float, target_u: float
) -> Tuple[float, float]:
    """
    Calculate azimuth and elevation from observer to target.
    
    Args:
        observer_e, observer_n, observer_u: Observer position in ENU (meters)
        target_e, target_n, target_u: Target position in ENU (meters)
    
    Returns:
        Tuple of (azimuth, elevation) in degrees
        - Azimuth: 0-360° from TRUE NORTH, clockwise
        - Elevation: -90 to +90° from horizon
    """
    # Vector from observer to target
    de = target_e - observer_e
    dn = target_n - observer_n
    du = target_u - observer_u
    
    # Horizontal distance
    horizontal_dist = math.sqrt(de * de + dn * dn)
    
    # Azimuth (from North, clockwise)
    azimuth = math.degrees(math.atan2(de, dn))
    if azimuth < 0:
        azimuth += 360.0
    
    # Elevation (from horizon)
    elevation = math.degrees(math.atan2(du, horizontal_dist))
    
    return (azimuth, elevation)


def distance_3d(
    x1: float, y1: float, z1: float,
    x2: float, y2: float, z2: float
) -> float:
    """Calculate 3D Euclidean distance."""
    return math.sqrt((x2 - x1)**2 + (y2 - y1)**2 + (z2 - z1)**2)


def haversine_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Calculate great-circle distance between two GPS points.
    
    Args:
        lat1, lon1: First point in degrees
        lat2, lon2: Second point in degrees
    
    Returns:
        Distance in meters
    """
    R = 6371000  # Earth radius in meters
    
    lat1_rad = math.radians(lat1)
    lat2_rad = math.radians(lat2)
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    
    a = (math.sin(dlat / 2) ** 2 + 
         math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(dlon / 2) ** 2)
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    
    return R * c

File: VR-UI/geo.ts
/**
 * WGS84 → ENU conversion, ported verbatim from code/math_utils/geo.py so the
 * headset agrees with the server about where things are. See the
 * geometry-conventions skill; verified against the Python round-trip example
 * there (37.7758,-122.4180,25 rel 37.7749,-122.4194,0 → E=123.339 N=99.894
 * U=24.998).
 *
 * Also owns the ENU → three.js scene mapping: three/WebXR is right-handed
 * Y-up with -Z forward, so (E, N, U) → (x=E, y=U, z=-N). With an unrotated
 * world group, -Z (the user's initial facing) is North.
 */

import { Vector3 } from './types';

// WGS84 ellipsoid parameters (identical to math_utils/geo.py)
export const WGS84_A = 6378137.0;
export const WGS84_B = 6356752.314245;
export const WGS84_E2 = 1.0 - (WGS84_B * WGS84_B) / (WGS84_A * WGS84_A);

const DEG = Math.PI / 180.0;

export function wgs84ToEcef(lat: number, lon: number, alt: number): Vector3 {
    const latRad = lat * DEG;
    const lonRad = lon * DEG;

    const sinLat = Math.sin(latRad);
    const cosLat = Math.cos(latRad);
    const sinLon = Math.sin(lonRad);
    const cosLon = Math.cos(lonRad);

    // Radius of curvature in the prime vertical
    const N = WGS84_A / Math.sqrt(1.0 - WGS84_E2 * sinLat * sinLat);

    const X = (N + alt) * cosLat * cosLon;
    const Y = (N + alt) * cosLat * sinLon;
    const Z = (N * (1.0 - WGS84_E2) + alt) * sinLat;

    return [X, Y, Z];
}

export function ecefToEnu(
    x: number, y: number, z: number,
    refLat: number, refLon: number, refAlt: number
): Vector3 {
    const [refX, refY, refZ] = wgs84ToEcef(refLat, refLon, refAlt);

    const dx = x - refX;
    const dy = y - refY;
    const dz = z - refZ;

    const latRad = refLat * DEG;
    const lonRad = refLon * DEG;

    const sinLat = Math.sin(latRad);
    const cosLat = Math.cos(latRad);
    const sinLon = Math.sin(lonRad);
    const cosLon = Math.cos(lonRad);

    const E = -sinLon * dx + cosLon * dy;
    const N = -sinLat * cosLon * dx - sinLat * sinLon * dy + cosLat * dz;
    const U = cosLat * cosLon * dx + cosLat * sinLon * dy + sinLat * dz;

    return [E, N, U];
}

export function wgs84ToEnu(
    lat: number, lon: number, alt: number,
    refLat: number, refLon: number, refAlt: number
): Vector3 {
    const [x, y, z] = wgs84ToEcef(lat, lon, alt);
    return ecefToEnu(x, y, z, refLat, refLon, refAlt);
}

// ---------------------------------------------------------------------------
// ENU ↔ three.js scene frame
// ---------------------------------------------------------------------------

/** ENU metres → scene-local coordinates (Y-up, North = -Z). */
export function enuToSceneXYZ(e: number, n: number, u: number): [number, number, number] {
    return [e, u, -n];
}

export function enuVecToSceneXYZ(v: Vector3): [number, number, number] {
    return [v[0], v[2], -v[1]];
}

/** Azimuth/elevation (deg, az 0=N CW, el 0=horizon up+) → unit ENU direction. */
export function anglesToEnuDirection(azimuthDeg: number, elevationDeg: number): Vector3 {
    const az = azimuthDeg * DEG;
    const el = elevationDeg * DEG;
    const cosEl = Math.cos(el);
    return [Math.sin(az) * cosEl, Math.cos(az) * cosEl, Math.sin(el)];
}
